### PR TITLE
Add new skin layer to allow components anchored to the playfield

### LINKED ISSUE #22388: Add per-ruleset skin layout customisation
Right now, a skin layout is saved globally for a skin. This doesn't work great for multiple rulesets, where the playfield may be attached to an edge of the screen and cause overlap. Also, in the new "argon" designs, UI elements have already been customised to often be attached to the playfield – this is also the case with classic skins like the osu!mania default.

This epic replaces existing requests for this:
- https://github.com/ppy/osu/issues/12871
- https://github.com/ppy/osu/issues/6124

This could be used to solve issues like:
- https://github.com/ppy/osu/issues/21949
- https://github.com/ppy/osu/issues/20422
- https://github.com/ppy/osu/issues/22362

It would both allow for applying per-ruleset defaults, and give users the option to further customise from those defaults. There should also likely be facilities in the editor to copy full layouts between rulesets (maybe not required if we have clipboard support).

### PR DESCRIPTION
**Pull request by peppy**
This has been a common request to cater for rulesets where the playfield changes in shape (ie. osu!mania with different column counts), or for cases where a user has applied playfield scale adjustments.

See commit message attached to 5bd06832d0a4e35bbbd52eea2ffb3970764e6d0e for the major change required to make this work.

https://github.com/ppy/osu/assets/191335/19be5e65-e7fe-4cc0-9e16-44b0ce18a0ee

Addresses #24379.
Closes https://github.com/ppy/osu/issues/22388

### CODE REVIEW

File: osu.Game/Overlays/SkinEditor/SkinComponentToolbox.cs [hunk 1]
```diff
@@ -23,14 +24,22 @@ public partial class SkinComponentToolbox : EditorSidebarSection
...
 
-        public SkinComponentToolbox(SkinComponentsContainer? target = null)
-            : base(target?.Lookup.Ruleset == null ? SkinEditorStrings.Components : LocalisableString.Interpolate($"{SkinEditorStrings.Components} ({target.Lookup.Ruleset.Name})"))
+        /// <summary>
+        /// Create a new component toolbox for the specified taget.
+        /// </summary>
+        /// <param name="target">The target. This is mainly used as a dependency source to find candidate components.</param>
+        /// <param name="ruleset">A ruleset to filter components by. If null, only components which are not ruleset-specific will be included.</param>
+        public SkinComponentToolbox(SkinComponentsContainer target, RulesetInfo? ruleset)
+            : base(ruleset == null ? SkinEditorStrings.Components : LocalisableString.Interpolate($"{SkinEditorStrings.Components} ({ruleset.Name})"))
```
Comment (alexpado) [inline osu.Game/Overlays/SkinEditor/SkinComponentToolbox.cs:39]: Unfortunatly, this breaks the test at `TestSceneSkinEditorComponentsList.TestToggleEditor()`

Comment (bdach) [reply osu.Game/Overlays/SkinEditor/SkinComponentToolbox.cs:39]: Fixed in aca8310cd1df87a131e1e1dfc7d0bc047101c578.

**Comment by bdach** [comment]
So I've fixed a few more test failures, but the last ones in `SkinDeserialisationTest` I'm not super sure what to do with. They're failing because the ZIP archives with test skins don't have `Playfield.json` inside, thus dying at

https://github.com/ppy/osu/blob/c21167e2ab3b6fd05e54d69537caeca13346c5db/osu.Game/IO/Archives/ZipArchiveReader.cs#L32-L34

But that is curious because skins from before this change continue to work, and I'm pretty sure this failure is _specific to reading skins directly from archives only_. The actual game will use a `DllResourceStore` or a `NativeStorage` underneath, that will just graciously return a null stream, and `Skin` will continue on its merry way.

The `FileNotFoundException` throw in `ZipArchiveReader` is 2016 code, so it's a bit scary to remove it... do we just add blank files to all of the zipped skins or what?

**Comment by peppy** [comment]
I'm willing to make a lunge and nuke that exception. It doesn't seem to be used anywhere and definitely doesn't match expectations.

**Review by bdach — APPROVED** [review]

**Pull request merged**